File: backend/app/schemas/pattern_engine.py
"""纹样基因引擎 Pydantic Schemas"""

from pydantic import BaseModel
from typing import Optional


class PatternGeneOut(BaseModel):
    """纹样基因输出"""
    id: int
    gene_id: str
    name: str
    shape_category: str
    meaning: str
    era: str = ""
    region: str = ""
    description: str = ""
    svg_viewbox: str = "0 0 100 100"
    svg_content: str = ""
    default_color: str = "#B8463A"
    tags: list[str] = []

    model_config = {"from_attributes": True}


class PatternMatchRequest(BaseModel):
    """基因匹配请求"""
    pattern_names: list[str]


class PatternMatchResponse(BaseModel):
    """基因匹配响应"""
    matched: list[PatternGeneOut]
    unmatched: list[str]  # 未匹配到的纹样名称


class GeneCultureResponse(BaseModel):
    """基因文化解读"""
    gene: PatternGeneOut
    cultural_meaning: str  # DeepSeek 生成的文化解读

File: frontend/src/services/patternEngine.ts
import api from './api'

export interface PatternGene {
  id: number
  gene_id: string
  name: string
  shape_category: string
  meaning: string
  era: string
  region: string
  description: string
  svg_viewbox: string
  svg_content: string
  default_color: string
  tags: string[]
}

export interface PatternMatchResponse {
  matched: PatternGene[]
  unmatched: string[]
}

export interface GeneCultureResponse {
  gene: PatternGene
  cultural_meaning: string
}

/** 提交纹样名称列表 → 匹配基因库 */
export async function matchPatterns(pattern_names: string[]): Promise<PatternMatchResponse> {
  const res = await api.post('/api/pattern-engine/match', { pattern_names })
  return res.data
}

/** 分页列出基因库 */
export async function listGenes(
  params?: { category?: string; meaning?: string; page?: number; page_size?: number }
): Promise<{ items: PatternGene[]; total: number }> {
  const res = await api.get('/api/pattern-engine/genes', { params })
  return res.data
}

/** 获取基因库分类列表 */
export async function listCategories(): Promise<{ categories: string[] }> {
  const res = await api.get('/api/pattern-engine/genes/categories')
  return res.data
}

/** 获取单个基因详情 + 文化解读 */
export async function getGeneDetail(gene_id: string): Promise<GeneCultureResponse> {
  const res = await api.get(`/api/pattern-engine/genes/${gene_id}`)
  return res.data
}
